File: dot_product/dot_product.py
import time
import random

ARRAY_SIZE = 10000

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

# Generate random float between 0 and 1
def randFloat():
    return random.random()

# Compute dot product for the array
def dotProduct(points):
    return sum(p.x * p.y for p in points)

# Generate a list of random points
def generateRandomPoints(size):
    return [Point(randFloat(), randFloat()) for _ in range(size)]

def main():
    timedMake = 0
    timedDot = 0

    for _ in range(100):
        start = time.time()
        points = generateRandomPoints(100000)
        timedMake += time.time() - start

        start = time.time()
        result = dotProduct(points)
        timedDot += time.time() - start

    print(f"Point create time taken: {timedMake:.5f} seconds")
    print(f"Dot product time taken: {timedDot:.5f} seconds")
    print(f"Total time taken: {timedDot+timedMake:.5f} seconds")

if __name__ == "__main__":
    main()

File: dot_product/dot_product.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define ARRAY_SIZE 10000

typedef struct {
    float x;
    float y;
} Point;

// Generate random float between 0 and 1
float randFloat() {
    return (float)rand() / RAND_MAX;
}

// Compute dot product for the array
float dotProduct(Point* arr, int size) {
    float sum = 0.0;
    for(int i = 0; i < size; i++) {
        sum += arr[i].x * arr[i].y;
    }
    return sum;
}

// Generate an array of random points on the heap
Point* generateRandomPoints(int size) {
    Point* points = (Point*)malloc(sizeof(Point) * size);
    if (!points) {
        fprintf(stderr, "Memory allocation failed!\n");
        exit(EXIT_FAILURE);
    }

    for(int i = 0; i < size; i++) {
        points[i].x = randFloat();
        points[i].y = randFloat();
    }
    return points;
}

int main() {


    // Measure performance
    clock_t timedMake = clock();
    clock_t timedDot = clock();
    for (int i=0;i<100;i++) {
        clock_t start = clock();
        Point *points = generateRandomPoints(100000);
        timedMake += clock()-start;
        start = clock();
        float result = dotProduct(points, ARRAY_SIZE);
        timedDot += clock()-start;
        free(points);
    }

    printf("Point create time taken: %f seconds\n", (double)timedMake/CLOCKS_PER_SEC);
    printf("Dot product time taken: %f seconds\n", (double)timedDot/CLOCKS_PER_SEC);
    printf("Total time taken: %f seconds\n", ((double)timedDot+(double)timedMake)/CLOCKS_PER_SEC);

    return 0;
}
